### DISCUSSION TITLE
How to save predicted video output?

### DISCUSSION THREAD

**Discussion by dlin95123** [ultralytics/ultralytics#5200] [Q&A]
When I use a jpg file as the source of predict, I can find a file in the runs folder and it has the image plus the bounding boxes and scores. However, if I use a mp4 file as the source, the file generated in the runs folder is an avi file of size 0. As an example, the following CLI command would generate the output image file:

`yolo detect predict model=yolov8s.pt source="c:\users\dlin1\downloads\street.jpg"`

However, if I change the source to a mp4 file, I don't get a playable output file. How do I fix that? Thanks.

**Comment by zZz-xg** [comment]
`from ultralytics import YOLO
model = YOLO("yolov8s.pt")    # pretrained YOLOv8n model
results = model.predict(source="c:\users\dlin1\downloads\street.jpg", # or: c:/users/dlin1/downloads/street.jpg ,# street.mp4 Or other format
                        show=True,   
                        device=0,   # GPU
                        save=True,      
                        save_txt=True
                        )   # predict on an image`

**Reply by zZz-xg** [reply]
yolo detect predict model=yolov8s.pt source='c:/users/dlin1/downloads/street.mp4' show=True,

**Reply by glenn-jocher** [reply]
@zZz-xg to save the predicted video output when using the `yolo detect predict` command, ensure that you have the correct dependencies installed and that your environment is properly set up. If you're encountering an issue where the output `.avi` file is of size 0, this could be due to various reasons such as codec issues, file permission problems, or an interruption during the video processing.

Make sure that:
- Your environment has the necessary codecs to write video files.
- The destination folder has write permissions.
- The source video is not corrupted and is in a readable format.

If the issue persists, you can try running the command with increased verbosity to get more detailed output logs, which might help in diagnosing the problem. Additionally, check the documentation at https://docs.ultralytics.com for any updates or common troubleshooting tips related to video output.

Remember, the `show=True` flag is used to display the results in a window, but it is not necessary for saving the output video. The command should automatically save the processed video in the `runs` directory without the need for additional flags. If you continue to face issues, please provide detailed error logs or messages, and we can further investigate the problem.

**Reply by XiaMinqi** [reply]
> @zZz-xg to save the predicted video output when using the `yolo detect predict` command, ensure that you have the correct dependencies installed and that your environment is properly set up. If you're encountering an issue where the output `.avi` file is of size 0, this could be due to various reasons such as codec issues, file permission problems, or an interruption during the video processing.
> 
> Make sure that:
> 
> * Your environment has the necessary codecs to write video files.
> * The destination folder has write permissions.
> * The source video is not corrupted and is in a readable format.
> 
> If the issue persists, you can try running the command with increased verbosity to get more detailed output logs, which might help in diagnosing the problem. Additionally, check the documentation at https://docs.ultralytics.com for any updates or common troubleshooting tips related to video output.
> 
> Remember, the `show=True` flag is used to display the results in a window, but it is not necessary for saving the output video. The command should automatically save the processed video in the `runs` directory without the need for additional flags. If you continue to face issues, please provide detailed error logs or messages, and we can further investigate the problem.

@glenn-jocher Hi, I'm wondering if it's possible to change the output format as mp4?

**Reply by glenn-jocher** [reply]
Currently, YOLOv8 saves video outputs in `.avi` format by default. To convert it to `.mp4`, you can use a tool like FFmpeg after the prediction process.

**Reply by XiaMinqi** [reply]
> Currently, YOLOv8 saves video outputs in `.avi` format by default. To convert it to `.mp4`, you can use a tool like FFmpeg after the prediction process.

Got it! Thanks a a lot!

**Reply by glenn-jocher** [reply]
You're welcome! If you have any more questions, feel free to ask.

**Reply by bayraktarberkay** [reply]
Saved videos have jpeg codec and this is not always the best option. Is there a parameter to change the codec format of the video?

**Reply by glenn-jocher** [reply]
@bayraktarberkay currently, Ultralytics YOLO uses default codecs based on the operating system. To change the codec, you can modify the source code in `ultralytics/engine/predictor.py` at the `save_predicted_images` method or use a tool like FFmpeg to re-encode the output video.

**Reply by martian7777** [reply]
@glenn-jocher isn't there any code base solution provided in yolo to convert into mp4 directly?

**Comment by mariam-162** [comment]
You can also change the format to mp4 using openCV..

```
def convert_avi_to_mp4(input_path, output_path):
    """Convert an AVI video file to MP4 format using OpenCV."""
    cap = cv2.VideoCapture(input_path)
    
    # Get the video's properties (frame width, height, and FPS)
    fps = cap.get(cv2.CAP_PROP_FPS)
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

    # Define the codec and create a VideoWriter object for MP4 output
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')  # Codec for MP4 format
    out = cv2.VideoWriter(output_path, fourcc, fps, (width, height))

    # Read and write each frame to the new video file
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        out.write(frame)
    
    # Release resources
    cap.release()
    out.release()
    print(f"Video saved as {output_path}")
```

**Reply by glenn-jocher** [reply]
@mariam-162 to save the predicted video output in a playable format, ensure the `save` argument is set to `True` in your command. If the output is still an AVI file, you can convert it to MP4 using OpenCV as shown in the provided code snippet.

**Reply by XiaMinqi** [reply]
Thanks for sharing! This is exactly the approach I'm currently using. 
However, as using OpenCV leads to more lines in code, I hope if an additional parameter could be set in this tool to control the output format in future version, then mp4 file could be generated gracefully.

**Reply by glenn-jocher** [reply]
Thank you for your feedback! We'll consider adding an option for output format control in future updates.
